crates/proto/src/xfer/dnssec_dns_handle/mod.rs

```diff
@@ -271,7 +279,8 @@ fn check_nsec(verified_message: DnsResponse, query: &Query) -> Result<DnsRespons
         }
     };
 
-    if !nsec_proof.is_secure() {
+    if !nsec_proof.is_secure() || (nsec_proof.is_secure() && query.query_type() == RecordType::DS) {
```

Comment:
This condition seems kinda non-intuitive, maybe add a little context in a comment?

Reply:
To address this (and the other comment about counter-intuitiveness,) I moved the Proof::Insecure decision into the verify_nsec and verify_nsec3 functions. 

verify_nsec and verify_nsec3 will now directly return Proof::Insecure for DS queries where the nsec/nsec3 validation code otherwise would have returned Proof::Secure (at least for scenarios where that makes sense - for example, in the no data response handling part of the NSEC3 code, but not in the nxdomain section.)